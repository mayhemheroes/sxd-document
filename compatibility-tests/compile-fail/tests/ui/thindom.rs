use sxd_document::Package;

fn cannot_mutate_connections_while_iterating_over_root_children() {
    let package = Package::new();
    let (s, mut c) = package.as_thin_document();

    let alpha = s.create_element("alpha");

    for _ in c.root_children() {
        c.append_root_child(alpha);
    }
}

fn cannot_mutate_connections_while_iterating_over_element_children() {
    let package = Package::new();
    let (s, mut c) = package.as_thin_document();

    let alpha = s.create_element("alpha");
    let beta = s.create_element("beta");

    for _ in c.element_children(alpha) {
        c.append_element_child(alpha, beta);
    }
}

fn cannot_mutate_connections_while_iterating_over_siblings() {
    let package = Package::new();
    let (s, mut c) = package.as_thin_document();

    let alpha = s.create_element("alpha");
    let beta = s.create_element("beta");

    for _ in c.element_preceding_siblings(alpha) {
        c.append_element_child(alpha, beta);
    }
}

fn nodes_cannot_live_outside_of_the_document() {
    let package = Package::new();

    let _node = {
        let (s, _) = package.as_thin_document();

        s.create_element("hello")
    };
}

fn main() {}
