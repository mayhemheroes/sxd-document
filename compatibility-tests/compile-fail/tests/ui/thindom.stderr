error[E0502]: cannot borrow `c` as mutable because it is also borrowed as immutable
  --> tests/ui/thindom.rs:10:9
   |
9  |     for _ in c.root_children() {
   |              -----------------
   |              |
   |              immutable borrow occurs here
   |              immutable borrow later used here
10 |         c.append_root_child(alpha);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here

error[E0502]: cannot borrow `c` as mutable because it is also borrowed as immutable
  --> tests/ui/thindom.rs:22:9
   |
21 |     for _ in c.element_children(alpha) {
   |              -------------------------
   |              |
   |              immutable borrow occurs here
   |              immutable borrow later used here
22 |         c.append_element_child(alpha, beta);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here

error[E0502]: cannot borrow `c` as mutable because it is also borrowed as immutable
  --> tests/ui/thindom.rs:34:9
   |
33 |     for _ in c.element_preceding_siblings(alpha) {
   |              -----------------------------------
   |              |
   |              immutable borrow occurs here
   |              immutable borrow later used here
34 |         c.append_element_child(alpha, beta);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here

error[E0597]: `s` does not live long enough
  --> tests/ui/thindom.rs:44:9
   |
41 |     let _node = {
   |         ----- borrow later stored here
42 |         let (s, _) = package.as_thin_document();
   |              - binding `s` declared here
43 |
44 |         s.create_element("hello")
   |         ^ borrowed value does not live long enough
45 |     };
   |     - `s` dropped here while still borrowed
